components/authz-service/server/v2/project.go

```diff
@@ -158,13 +164,47 @@ func (s *state) UpdateProject(ctx context.Context,
 }
 
 func (s *state) ApplyRulesStart(
-	context.Context, *api.ApplyRulesStartReq) (*api.ApplyRulesStartResp, error) {
-	s.log.Info("apply project rules: START")
-	err := s.projectUpdateManager.Start()
+	ctx context.Context, _ *api.ApplyRulesStartReq) (*api.ApplyRulesStartResp, error) {
+	// NOTE (tc): Only one call to ApplyRulesStart can happen at a time.
+	// This should be good enough to prevent race conditions for single node,
+	// in conjunction with the table locking that happens in store.ApplyStagedRules.
+	// Can still get into a weird state if we panic, but a re-apply will fix things up.
+	// We will be refactoring for multi-node using workflow tooling in the near future.
+	s.applyRuleMux.Lock()
+	defer s.applyRuleMux.Unlock()
+
+	switch s.projectUpdateManager.State() {
+	case config.NotRunningState:
+		break
+	case config.RunningState:
+		return nil, status.Error(codes.FailedPrecondition,
+			"cannot apply rules: apply already in progress")
+	default:
+		return nil, status.Error(codes.Internal,
+			"failed to parse apply state")
+	}
+
+	err := s.store.ApplyStagedRules(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal,
+			"error applying staged projects: %s", err.Error())
+	}
+
+	// TODO (tc): If we panic between here and manager Start, we will be in a state where the rules
+	// have been updated in the database but Refresh has not been kicked off.
+	// We will be refactoring with workflow to make this safer soon.
+	err = s.policyRefresher.Refresh(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal,
+			"error refreshing policy cache: %s\nthe rules were updated but the apply was not started, please try again", err.Error())
```

Comment:
[nit] Newlines? Let's not do that. Let's log if we need more info, but multi-line error messages are weird. 😉